kolibri/core/assets/src/api_resource.js

```diff
@@ -281,12 +437,16 @@ class Resource {
     if (!(model instanceof Model)) {
       return this.addModelData(model);
     }
-    if (!this.models[model.id]) {
-      this.models[model.id] = model;
-    } else {
-      this.models[model.id].set(model.attributes);
+    // Don't add to the model cache if the id is not defined.
+    if (model.id) {
+      if (!this.models[model.id]) {
+        this.models[model.id] = model;
+      } else {
+        this.models[model.id].set(model.attributes);
+      }
+      return this.models[model.id];
     }
-    return this.models[model.id];
+    return model;
   }
 
```

Comment:
Here we should add a `removeModel` method, that removes the model passed in from `this.models`. This will clean up extant references to the model, and allow it to be garbage collected.